### ISSUE TITLE
Issue with .net core 2.0 upgrade.

### ISSUE THREAD

**Issue by asadsahi** [openiddict/openiddict-core#424]
Hi @PinpointTownes I am trying to upgrade [my project](https://github.com/asadsahi/AspNetCoreSpa) to .net core 2.0. Branch name is **netcore2**.

Managed to login using credentials and token model is returned from server, including refresh_token. But when request is sent using refresh token (e.g by doing full page reload), getting following error:

`
{"message":"No authentication handler is configured to authenticate for the scheme: ASOS","isError":true,"detail":"   at Microsoft.AspNetCore.Http.Authentication.Internal.DefaultAuthenticationManager.<GetAuthenticateInfoAsync>d__12.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at AspNetCoreSpa.Server.Controllers.api.AuthorizationController.<Exchange>d__4.MoveNext() in c:\\Users\\asahi\\AspNetCoreSpa\\Server\\Controllers\\api\\AuthorizationController.cs:line 115\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionMethodAsync>d__12.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextActionFilterAsync>d__10.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeInnerFilterAsync>d__14.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextExceptionFilterAsync>d__23.MoveNext()","errors":null}`

Here is the [line ](https://github.com/asadsahi/AspNetCoreSpa/blob/master/Server/Extensions/ServiceCollectionExtensions.cs#L95 )where I have registered in service collection. 

Anything obvious missing? 

Thanks

**Comment by kevinchalet** [comment]
Hey @asadsahi,

> Anything obvious missing?

It looks like you're still using the old authentication manager that was obsoleted in 2.0 (and no longer works). I'd recommend reading https://github.com/aspnet/Announcements/issues/262 to determine how you can migrate your own code to the new authentication service.

**Comment by kevinchalet** [comment]
Please also note that the aspnet-contrib OAuth2 social providers for 1.0 won't work with 2.0 (https://github.com/asadsahi/AspNetCoreSpa/blob/netcore2/AspNetCoreSpa.csproj#L13-L14). Unfortunately, there's currently no version compatible with 2.0 so you should maybe exclude them until a compatible version is released.

**Comment by asadsahi** [comment]
Thanks @PinpointTownes . I followed most of the code from the example code of this project.

But if you reckon that its something to do with .net core code rather openidconnect. I will try to read and understand what it could be. Changes seem to be quite daunting as there are so many.

Also, the those social providers aren't working so their registration is commented out anyway so I will keep and eye for the update.

**Comment by asadsahi** [comment]
@PinpointTownes 

Also, any plans to upgrade https://github.com/openiddict/openiddict-samples to .net core 2?. Followed those sample initially when i implement openiddict in my project. so They will be quite handy too.

Thanks

**Comment by kevinchalet** [comment]
> Also, any plans to upgrade https://github.com/openiddict/openiddict-samples to .net core 2?

Not before 2.0 is officially released as the new LTS version (which may not happen before 2.0.1 or 2.0.2).

**Comment by kevinchalet** [comment]
> Changes seem to be quite daunting as there are so many.

@HaoK is there a discussion topic where folks can ask questions about the "auth 2.0" changes? I didn't see one when reading the announcement.

**Comment by HaoK** [comment]
Good idea, I'll create one now and link it to the announcement

**Comment by JanEggers** [comment]
@PinpointTownes  hi there .net core 2.0 is final and im having problems getting password / refreshtoken flow working. updated samples would be very helpful

**Comment by kevinchalet** [comment]
@JanEggers it's on my todo list :sweat_smile: 

Can you tell us more about the issue you're seeing?

**Comment by JanEggers** [comment]
my biggest problem is in AuthorizationController:
1. error CS7069: Reference to type 'AuthenticationTicket' claims it is defined in 'Microsoft.AspNetCore.Authentication', but it could not be found

2. Authentication.GetAuthenticateInfoAsync should be replaced with GetTokenAsync according to the migration announcement. that function just returns a string and so i dont know how to get principal and properties to create the ticket.

**Comment by JanEggers** [comment]
by the way it would be nice if that functions in the samples could be refactored into a service that could be used. right now i copy and paste the code from the samples to all projects im using your library...

**Comment by kevinchalet** [comment]
> error CS7069: Reference to type 'AuthenticationTicket' claims it is defined in 'Microsoft.AspNetCore.Authentication', but it could not be found

`AuthenticationTicket` is now part of `Microsoft.AspNetCore.Authentication.Abstractions` (namespace: `Microsoft.AspNetCore.Authentication`).

> Authentication.GetAuthenticateInfoAsync should be replaced with GetTokenAsync according to the migration announcement.

I did some CTRL+F and couldn't find a trace of `GetTokenAsync()` in the announcement, so not sure where you saw that. `GetAuthenticateInfoAsync()` was replaced by `AuthenticateAsync()`.

**Comment by kevinchalet** [comment]
> by the way it would be nice if that functions in the samples could be refactored into a service that could be used. right now i copy and paste the code from the samples to all projects im using your library...

It's not something I plan to do in the short term. For full flexibility, we opted for a template approach exactly like what Identity does with its `AccountController`/`ManageController`, which sounds better than forcing you to use our own authentication logic.

**Comment by JanEggers** [comment]
yet another issue on startup:

the type "Microsoft.AspNetCore.Builder.AuthenticationOptions" in the Assembly "Microsoft.AspNetCore.Authentication, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" could not be loaded

   at Microsoft.AspNetCore.Builder.OpenIddictExtensions.EnableTokenEndpoint(OpenIddictBuilder builder, PathString path)

**Comment by kevinchalet** [comment]
Are you sure you're using the OpenIddict 2.0 packages?

**Comment by JanEggers** [comment]
oops no sry

**Comment by ghost** [comment]
@JanEggers did you manage to solve your problem?

**Comment by JanEggers** [comment]
the startup error is gone by updating 2.0 preview1 packages but i still dont know how to correctly implement AccountController CreateAuthTicket

**Comment by kevinchalet** [comment]
@JanEggers I'm a bit surprised you're blocked by that since there's no difference between 1.x and 2.x except the `GetAuthenticateInfoAsync()` -> `AuthenticateAsync()` change.

**Comment by ghost** [comment]
@JanEggers total newb here I also have the same startup errors as you and I updated the packages but I still get the errors. Is there something wrong with my version? `<PackageReference Include="OpenIddict" Version="2.0.0-preview1" />
    <PackageReference Include="OpenIddict.EntityFrameworkCore" Version="2.0.0-preview1" />
    <PackageReference Include="OpenIddict.Mvc" Version="2.0.0-preview1" />`

**Comment by kevinchalet** [comment]
@vincealdrin there's no final `2.0.0-preview1` version. Use wildcard versions: `2.0.0-preview1-*` or `2.0.0-*`

**Comment by JanEggers** [comment]
@PinpointTownes: i tried AuthenticateAsync but that fails for valid refreshtokens. is there any way to diagnose why it fails?

succeeded is false and none is true all other properties of info are null.

**Comment by kevinchalet** [comment]
@JanEggers enable logging and take a look at your logs?

FYI, I ported all the OpenIddict samples to 2.0 last night, including the ROPC and refresh flow samples: https://github.com/openiddict/openiddict-samples

**Comment by asadsahi** [comment]
Thanks @PinpointTownes .

**Comment by JanEggers** [comment]
hi @PinpointTownes ,

sorry to bother you but i cant get it working.

i updated my playground project and im having the same issues, i digged a little further. the login / refresh requests seem to be ok but the handling of the auth token itself seems not to work.

i copied over the logs at 

https://gist.github.com/JanEggers/f5014ab1ca21d2846b6eab550f8f554c

and updated the code at

https://github.com/JanEggers/Playground

for me it looks like cookie auth is used and not OAuth, but i dont see any authentification logs just authorization that fails and the challange after that.

**Comment by JanEggers** [comment]
finally i found my error.

i called AddIdentity after AddOAuthValidation which overrides default authentification and challange themes to use cookie auth. narf that wasted 2 days of my life.

**Comment by kevinchalet** [comment]
@JanEggers glad you fixed your issue!

Identity is indeed a bit greedy and if you try to register your default scheme before calling Identity, it will be ignored.

@HaoK would that be worth considering updating `services.AddIdentity()` to avoid overriding the default schemes when they are not null (i.e when already set by the developer?).

**Comment by HaoK** [comment]
Sure, that's something we can consider, please file an issue in identity.  But I'm not sure we can do it in 2.1 as that might be considered a breaking change (even if its a 'good' one in most cases).

**Comment by kevinchalet** [comment]
@HaoK thanks! Done: https://github.com/aspnet/Identity/issues/1381.

**Comment by adnan-kamili** [comment]
Hi @PinpointTownes ,

After upgrading to .net core 2.0, authorization is failing in JWT, in logs I can see user is (null) causing the challege to get executed:

https://github.com/adnan-kamili/AspNetCore-Web-Api-Rest-Starter-Kit/blob/master/Startup.cs

**Comment by adnan-kamili** [comment]
To get it working I had to explicitly set:
```
services.AddAuthentication(options => {
                options.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
                options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
            })
```

**Comment by Flash717** [comment]
@adnan-kamili thanks your posts, ran into same issue migrating to .net core 2.0, setting default-schemes as per your post resolved our token validation issue.

**Comment by kevinchalet** [comment]
Closing, as I believe all/most questions have been answered.

@asadsahi let us know if you're still blocked.

**Comment by MovGP0** [comment]
Instead of `HttpContext.Authentication.GetAuthenticateInfoAsync(…)` you need to use `HttpContext.AuthenticateAsync(…)`.
